web3_proxy/src/frontend/authorization.rs

```diff
@@ -1084,6 +1098,39 @@ impl Web3ProxyApp {
         }
     }
 
+    /// Get the balance for the user.
+    ///
+    /// If a subuser calls this function, the subuser needs to have first attained the user_id that the rpc key belongs to.
+    /// This function should be called anywhere where balance is required (i.e. only rpc calls, I believe ... non-rpc calls don't really require balance)
+    pub(crate) async fn balance_checks(&self, user_id: u64) -> Web3ProxyResult<Arc<AtomicF64>> {
+        if user_id == 0 {
+            return Ok(Arc::new(AtomicF64::from(0.)));
+        }
+        match NonZeroU64::try_from(user_id) {
+            Err(_) => Ok(Arc::new(AtomicF64::from(0.))),
+            Ok(x) => {
+                self.user_balance_cache
+                    .get_or_insert_async(&x, async move {
+                        let db_replica = self
+                            .db_replica()
+                            .web3_context("Getting database connection")?;
+
+                        let balance: Decimal = balance::Entity::find()
+                            .filter(balance::Column::UserId.eq(user_id))
+                            .one(db_replica.conn())
+                            .await?
+                            .context("did not find related balance")?
+                            .available_balance;
+
+                        Ok(Arc::new(AtomicF64::from(balance.to_f64().context(
+                            "Balance is too high, or too low to turn into a float!",
```

Comment:
as mentioned above, lets try an `Arc<tokio::sync::RwLock<Decimal>>` for now

Reply:
implementing this